maitake/src/task/storage.rs

```diff
@@ -0,0 +1,74 @@
+use super::Task;
+use core::future::Future;
+use core::ptr::NonNull;
+
+/// A trait representing the storage medium of a [`Task`].
+///
+/// This is used to contain tasks at runtime, and abstract over the
+/// type erasure and type recovery steps, e.g. converting a heap
+/// allocation type into a NonNull pointer, and recovering it
+/// back into a heap allocation from a NonNull pointer
+///
+/// This trait is exposed publicly to allow for end users to implement
+/// it in their own heap allocation types, if they are not specifically
+/// a [`Box`].
+///
+/// This trait is ONLY appropriate for heap allocation types that represent
+/// exclusive ownership of the contained data, as it may be mutated while
+/// in pointer form. For example, the `Box` type would match this guarantee,
+/// but the `Arc` type would not, as it allows for shared access, and NOT
+/// exclusive mutable access.
+///
+/// NOTE: The type that implements this trait is typically NOT the heap
+/// allocation type itself, but a "Marker Type" that represents the
+/// intended storage medium. See the [`BoxStorage`] type (available with the
+/// "alloc" feature active) for an implementation example
+///
+/// [`Task`]: crate::task::Task
+/// [`Box`]: alloc::boxed::Box
+/// [`BoxStorage`]: crate::task::storage::BoxStorage
+pub trait Storage<S, F: Future>: Sized {
+    /// The type of a stored Task.
+    ///
+    /// As the type that implements the Storage trait is a Marker Type,
+    /// This associated type is the actual heap type that will be used
+    /// to contain the Task.
+    type StoredTask;
+
+    /// Convert an owned, heap allocated Task type to a raw pointer
+    ///
+    /// This method should produce a NonNull pointer, while not actually
+    /// dropping the contained task.
+    fn into_raw(task: Self::StoredTask) -> NonNull<Task<S, F, Self>>;
+
+    /// Convert a raw task pointer into an owned, heap allocated Task type
+    ///
+    /// This method should produce a heap allocated type, which can be
+    /// dropped to perform the correct destructor actions
+    fn from_raw(ptr: NonNull<Task<S, F, Self>>) -> Self::StoredTask;
+}
+
+feature! {
+    #![feature = "alloc"]
+    use alloc::boxed::Box;
+
+    /// A type representing [`Box`] storage of a task
+    ///
+    /// [`Box`]: alloc::boxed::Box
+    pub struct BoxStorage;
```

Comment:
@hawkw should I add an internal private type to make this externally unconstructable?

I don't think you can *do* anything with the type anyway.